### ISSUE TITLE
OSS to expose classic optimizer compliant param_groups

### ISSUE THREAD

**Issue by blefaudeux** [facebookresearch/fairscale#28]
## 🚀 Feature
Expose the "classic" param_groups from the OSS optimizer wrapper, either via a new API (not as nice) or if possible via the classic .param_groups attribute (get and set) 

## Motivation
Some frameworks like classy vision use that to alter the LR per layer or over time (or both). Exposing this would make OSS integration a lot easier

## Pitch
Either use dedicated getter and setter properties which change behaviour depending on whether the call comes from the outside or the super constructor, or add a specific API to get the consolidated param_groups and set the consolidated param_groups

## Alternatives
- Destroy the optimizer every time there's an adjustment
- Only expose adjustments which apply to the whole param_groups ("set all LR to a given value")

**Comment by blefaudeux** [comment]
cc @mannatsingh @vreis , FYI, I'm trying to think of an elegant way to handle that. The crux of the problem right now with Classy integration is that ClassyOptimizer is a wrapper, OSS also is a wrapper, and the LR schedulers expect to interact with 'self.optimizer.param_groups'. This is unambiguous for a "normal" optimizer (there's a single state in flight at any time), but it becomes ambiguous when the state is sharded because it could be the replica's param_groups or the whole optimizer param_groups (which is virtual at this point). Classy assumes the latter, OSS silently exposes the former.

**Comment by mannatsingh** [comment]
Yeah, that's a good point. At the same time, I think we can let Classy think it's overriding the whole params when in fact it would just be overriding the sharded params. I don't think Classy actually *needs* the params to be the complete set. At the end of the day, what Classy is doing is taking some param groups and splitting them into subsets to configure different optimizer settings. But I am not sure how involved this is on the OSS side.

**Comment by blefaudeux** [comment]
@mannatsingh there are two possible practical issues I think, at least that's what I'm seeing in some of the unit test suite:
- Classy pulls a rank state and alters some keys, only to find that at times they are not there (->`KeyError`), if `param_groups` only exposes a shard
- Classy pulls the whole state and "pushes" the new one, but only the reference replica is affected, if `param_groups` exposes the whole virtual state but does not sync back

**Comment by mannatsingh** [comment]
> Classy pulls a rank state and alters some keys, only to find that at times they are not there (->KeyError), if param_groups only exposes a shard

Hmm - can the shards actually change? Otherwise I'm not sure why the keys can go missing?

> Classy pulls the whole state and "pushes" the new one, but only the reference replica is affected, if param_groups exposes the whole virtual state but does not sync back

Right now Classy should ideally push the new state on all replicas, but yeah, this can be tricky. On Classy's side if we show Classy the whole virtual state, Classy should act on it the same way on all replicas. And then the work is offloaded to the virtualization logic to actually deal with this properly.

**Comment by blefaudeux** [comment]
> > Classy pulls a rank state and alters some keys, only to find that at times they are not there (->KeyError), if param_groups only exposes a shard
> 
> Hmm - can the shards actually change? Otherwise I'm not sure why the keys can go missing?
I mean that you can have something [like this ](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.LambdaLR), where you have one scheduler per param_group for instance. Now if everything is done from the reference rank, then it does not have all the groups in its own `.param_groups` attribute

**Comment by mannatsingh** [comment]
Actually, so that it's easier for me to understand, do you mind writing a simple example on how parameter groups are set in OSS right now across different ranks?

**Comment by blefaudeux** [comment]
Right now, it looks like (pseudo code)

`zero = OSS(model_params)
zero.param_groups : model_params # through the torch.optimizer default constructor, so this has nothing specific to the wrapped optimizer, no LR for instance
zero.optim.param_groups :  rank_and_optim_specific_param_groups # this is probably what classy is looking for, but rank specific`

**Comment by mannatsingh** [comment]
Thanks @blefaudeux ! What I was actually asking was slightly different -
```
param_groups = [{"params": trunk_params, "lr": 0.1}, {"params": head_params, "lr": 0.2}]

optimizer = SGD(param_groups)

# is the same supported for OSS?
zero = OSS(param_groups)
```

**Comment by blefaudeux** [comment]
> Thanks @blefaudeux ! What I was actually asking was slightly different -
> 
> ```
> param_groups = [{"params": trunk_params, "lr": 0.1}, {"params": head_params, "lr": 0.2}]
> 
> optimizer = SGD(param_groups)
> 
> # is the same supported for OSS?
> zero = OSS(param_groups)
> ```

yes, the same is supported by OSS !

**Comment by mannatsingh** [comment]
Ah, now I'm getting what the issue is. Sorry for being daft!
Just to summarize, here's what's happening -

```python
param_groups = [{"params": trunk_params, "lr": 0.1}, {"params": head_params, "lr": 0.2}]
classy_oss = ClassyOSS()
classy_oss.set_param_groups(param_groups) # this calls OSS(param_groups), which works

classy_oss.step() # this needs self.optimizer.param_groups to update the dicts, but that isn't available  
```

**Comment by blefaudeux** [comment]
> Ah, now I'm getting what the issue is. Sorry for being daft!
> Just to summarize, here's what's happening -
> 
> ```python
> param_groups = [{"params": trunk_params, "lr": 0.1}, {"params": head_params, "lr": 0.2}]
> classy_oss = ClassyOSS()
> classy_oss.set_param_groups(param_groups) # this calls OSS(param_groups), which works
> 
> classy_oss.step() # this needs self.optimizer.param_groups to update the dicts, but that isn't available  
> ```

my fault, I was not clear enough.. It's basically a case of russian dolls exposing similar but not identical attributes, and outside accessors who are misled. As shown in a diff this can be "fixed" on the client side, but it's not very elegant

**Comment by blefaudeux** [comment]
Turns out that there were a couple of smallish issues, one which does not really have consequences (OSS is a wrapper, but disguised in an optimizer itself. Classy expects all the optimizer attributes to be at this level while they only are true of the shard -this is the only one which steps-), and another one which was a botched state restoration, fixed in the attached PR.
